src/pkcs11/framework-pkcs15.c

```diff
@@ -3680,6 +3680,58 @@ pkcs15_prkey_get_attribute(struct sc_pkcs11_session *session,
 }
 
 
+
+static CK_RV
+pkcs15_prkey_check_pss_param(CK_MECHANISM_PTR pMechanism, CK_ULONG hlen)
+{
+	CK_RSA_PKCS_PSS_PARAMS *pss_param;
+
+	if (pMechanism->pParameter == NULL)
+		return CKR_OK;				// Support applications that don't provide CK_RSA_PKCS_PSS_PARAMS
+
+	if (pMechanism->ulParameterLen != sizeof(CK_RSA_PKCS_PSS_PARAMS))
+		return CKR_MECHANISM_PARAM_INVALID;
+
+	pss_param = (CK_RSA_PKCS_PSS_PARAMS *)pMechanism->pParameter;
+
+	// Hash parameter must match mechanisms or length of data supplied for CKM_RSA_PKCS_PSS
+	switch(pss_param->hashAlg) {
+	case CKM_SHA_1:
+		if (hlen != 20)
+			return CKR_MECHANISM_PARAM_INVALID;
+		break;
+	case CKM_SHA256:
+		if (hlen != 32)
+			return CKR_MECHANISM_PARAM_INVALID;
+		break;
+	default:
+		return CKR_MECHANISM_PARAM_INVALID;
+	}
+
+	// SmartCards typically only support MGFs based on the same hash as the
+	// message digest
+	switch(pss_param->mgf) {
+	case CKG_MGF1_SHA1:
+		if (hlen != 20)
+			return CKR_MECHANISM_PARAM_INVALID;
+		break;
+	case CKG_MGF1_SHA256:
+		if (hlen != 32)
+			return CKR_MECHANISM_PARAM_INVALID;
+		break;
+	default:
+		return CKR_MECHANISM_PARAM_INVALID;
+	}
+
+	// SmartCards typically support only a salt length equal to the hash length
```

Comment:
The same for this condition -- PKCS#11 happily allows wider range of combinations.

Reply:
And some cards/tokens (like YubiHSM2) do support other SHA2 family members.